### PR TITLE
preparations for cv merge to psi-ms

### PR DESCRIPTION
**Pull request by mwalzer**
see #141

### CODE REVIEW

File: cv/qc-cv.obo [hunk 1]
```diff
@@ -42,336 +40,330 @@ name: has_tuple
 id: has_units
 name: has_units
 
+
+[Term]
+id: MS:4000000
+name: PSI-MS CV Quality Control Vocabulary
+def: "PSI-MS CV Quality Control Vocabulary." [PSI:MS]
+is_a: MS:0000000 ! Proteomics Standards Initiative Mass Spectrometry Vocabularies
```
Comment (timosachsenberg) [inline cv/qc-cv.obo:48]: minor observation: so far never used with is_a
https://github.com/HUPO-PSI/psi-ms-CV/blob/master/psi-ms.obo#L9565

Comment (mwalzer) [reply cv/qc-cv.obo:48]: Thanks, good point. We'll have to fix that.

**Comment by mwalzer** [comment]
@julianu @bittremieux @cbielow Please let me know if you think we can use the PR as it is for a merge request with PSI-MS-CV or not and where you think we need additional structure or better names/definitions.

**Comment by nilshoffmann** [comment]
@mwalzer I prepared pull request #159 against this one which resolves some remaining QC prefixes and removes and extra id

**Comment by sneumann** [comment]
Hi, I am doing some stupid things here: 
```
cut -d : -f 1 qc-cv.obo  | sort | uniq -c | sort -n 
...
    234 [Term]
    237 def
    246 id
    246 name
    602 is_a
```
and trying to see why we have more `Term`s than `def`s . Found a few issues:

1) id: MS:4000107 has two [Term]s
1) QC:4000027 and QC:4000028 have no def
1) five trailing spaces (don't think they really matter)
```
grep -n '^\[Term\] $' qc-cv.obo # line numbers
2193:[Term] 
2265:[Term] 
2281:[Term] 
2297:[Term] 
2303:[Term] 
```
I see several CV issues https://github.com/HUPO-PSI/mzQC/issues?q=is%3Aissue+is%3Aopen+label%3ACV but several of them are adressed in commits by Matthias. Can they be closed ? 
I focused on such technical things, little content review done. 
Yours, Steffen

**Comment by bittremieux** [comment]
Great, thanks for checking those issues @sneumann. I think I've fixed them (with a small git snafu in the mix).

Most of the open CV issues can indeed be closed when this branch is merged in.

**Review by bittremieux — COMMENTED** [review]
Some comments/questions:

- [ ] Should the namespace coverage be `sequence(7,4000000,4999999)` instead?
- [ ] `has_relation` is a weird name for the typedef. It doesn't exist in PSI-MS-CV yet. Can we bring it in line with that CV or rename it to something better?
- [ ] `has_type` should be renamed to `has_value_type` for consistency with PSI-MS-CV.
- [ ] `has_tuple` is not used and can be removed.
- [ ] "ID free" etc. is a metric category imo, and should be specified using `relationship` rather than `is_a`.
- [ ] A bunch of metrics are missing unit and type specifications that could be included.
- [ ] `MS:4000055` and `MS:4000056`: The name uses "quantile", but the definition details that it's quartiles.

The CV is too unwieldy to fully go through it. At least I have not been able to do so.
Unfortunately I think that the CV would benefit from carefully reviewing all QC terms individually with a checklist in hand, because several terms that I looked at so far still have issues:
- Is the metric properly defined?
- Have all potential relationships and categorizations been specified?
- Has the unit been specified?
- Has the value type been specified?
- Have links to other terms been listed?
- Has the term been checked for duplication in other CVs?

**Comment by sneumann** [comment]
Regarding "Is the metric properly defined?" @tnaake could comment on the ones he found most difficult to understand while implementing them. Yours, Steffen

**Comment by tnaake** [comment]
Hi all and many thanks for the fantastic conceptualization of MS QC metrics,

I am currently implementing together with @jorainer parts of the mzQC metrics in `R` (see https://github.com/tnaake/MsQuality for the repo). 

We are currently focusing on the metrics that can be calculated by the usage of `Spectra` (https://github.com/rformassspectrometry/Spectra) and `MsExperiment` (https://github.com/rformassspectrometry/MsExperiment) objects as these are the central architecture to store MS data in R. 

Looking at some of the metrics, at least I was a bit puzzled how to calculate some of them. It would be great if you could comment if we (I) interpreted those correctly (and possibly adjust the definitions for easier understanding). Please note, that for some of the metrics we currently deviate from the definitions. Some definitions e.g. leave it open if a certain quantile needs to be calculated, while we focus for some functions on the `0`, `.25`, .5`, `.75`. 1` quantile. Please also note, that we are dealing here sometimes explicitly with `NA` values. When dealing with real-world data sets, we expect that there are no `NA`s. 

 - RT over TIC quantile (QC:4000054): 
     -  (1) the `Spectra` object is ordered according to the retention time,
     -  (2) the cumulative sum of the ion count is calculated (TIC),
     -  (3) the quantiles are calculated according to the given quantiles, e.g. the 0%, 25%, 50%, 75% and 100% quantile,
    - (4) the retention time/relative retention time (retention time divided by the total run time taking into account the minimum retention time) is calculated,
    -  (5) the (relative) duration of the LC run after which the cumulative TIC exceeds (for the first time) the respective quantile of the cumulative TIC is calculated and returned.

- MS1/MS2 quantiles RT fraction (QC:4000055/QC:4000056):
    - (1) the retention time duration of the whole `Spectra` object is determined (taking into account all the MS levels),
    - (2) the `Spectra` object is filtered according to the MS level and subsequently ordered according to the retention time
    - (3) the MS events are split into four (approximately) equal parts,
    - (4) the relative retention time is calculated (using the retention time duration from (1) and taking into account the minimum retention time),
    - (5) the relative retention time values associated to the MS event parts are returned.

- MS1 quantile TIC change ratio to Quantile 1 (QC:4000057) or to previous quantile (QC:4000058):
    - *TIC changes* are interpreted as follows:
        - (1) the cumulative sum (`cumsum`) of the  `spectra`'s TIC is calculated  (with spectra ordered by retention time),
        - (2) quartiles are then calculated on these, 
        - (3) for *QC:4000057* the log2 ratio between the 25, 50, 75 and 100% quantile to the 0% quantile is calculated. For *QC:4000058*  ratios between the 25/0, 50/25, 75/50 and 100/75% quantiles are calculated (this function interprets the *quantiles* from the [PSI:QC] definition as *quartiles*, i.e. the 0, 25, 50, 75 and 100% quantiles are used),
        - The `log2` values are returned instead of the `log` values.

- Peptide identification rate of the interquartile RT period (QC:4000073):
    - (1) the `Spectra` object is filtered according to the MS level,
    - (2) the retention time values are obtained,
    - (3) the 25% and 75% quantiles are obtained from the retention time values (`NA` values are removed),
    - (4) the number of eluted features between this 25% and 75% quantile is calculated,
    - (5) the number of features is divided by the interquartile range of the retention time and returned.

- Area under TIC RT quantiles (QC:4000078):
    - (1) the `Spectra` object is filtered according to the MS level, 
    - (2) the `Spectra` object is ordered according to the retention time, 
    - (3) the 0%, 25%, 50%, 75%, and 100% quantiles of the retention time values are obtained, 
    - (4) the ion count of the intervals between the 0%/25%, 25%/50%, 50%/75%, and 75%/100% are obtained (this function interprets the *quantiles* from the [PSI:QC] definition as *quartiles*, i.e. the 0, 25, 50, 75 and 100% quantiles are used),
    - (5) the ion counts of the intervals are summed (TIC) and the values returned.

- Median of TIC values in the RT range in which the middle half of peptides are identified (QC:4000130):
    - (1) the `Spectra` object is filtered according to the MS level, 
    - (2) the `Spectra` object is ordered according to the retention time,
    - (3) the features between the 1st and 3rd quartile are obtained (half of the features that are present in `spectra`),
    - (4) the ion count of the features within the 1st and 3rd quartile is obtained,
    - (5) the median value of the ion count is calculated (`NA` values are removed) and the median value is returned.

- Median of TIC values in the shortest RT range in which half of the peptides are identified (QC:4000132):
    - (1) the `Spectra` object is filtered according to the MS level,
    - (2) the `Spectra` object is ordered according to the retention time, 
    - (3) the number of features in `spectra` is obtained and the number for half of the features is calculated,
    - (4) iterate through the features (always by taking the neighbouring half of features) and calculate the retention time range of the set of features,
    - (5) retrieve the set of features with the minimum retention time range,
    - (6) calculate from the set of (5) the median TIC (`NA` values are removed) and return it.

Many thanks, 
@tnaake

**Comment by bittremieux** [comment]
Very nice, I'm excited to see this work!

I interpret the metrics the same way as you do, although I have not implemented any of these myself.

The reference implementation for the first three metrics is QuaMeter IDfree. QuaMeter IDfree is straightforward to get working, so you should be able to compare your results against theirs relatively easily.

The final four metrics are from NIST MSQC I think. I would not recommend trying to get NIST MSQC working, if that even still exists. But these metrics are also implemented in QuaMeter IDbased mode. It takes a bit more effort to get those results because you need to have identification data as well, so ymmv with comparing against that.

Pinging @dtabb73, who has supervised development of QuaMeter.

**Comment by tnaake** [comment]
Dear @bittremieux 

many thanks for checking your interpretation of the metrics with ours. We also hope that the package will be used within the R community heavily as it represents an interface with the `Spectra` and `MsExperiment` and the concepts of mzQC and are looking forward to getting some end user feedback after releasing the package. 

We will have a look at QuaMeter IDfree and will check if we get the same results. Thanks for the hint!

Best wishes,
@tnaake

**Comment by mwalzer** [comment]
Hi all,
as @bittremieux said: very unwieldy. I think to help progress, we'll merge this one to get the main branch updated with the progress made in this PR, then move in smaller steps (and more focused PRs). I'd suggest we prepare thematic stripes of terms to go into PSI-MS first to keep the size manageable, using the checklist above. 

@tnaake @sneumann Also very excited about your work! I'll `@ mention` in the future PRs that relate to your comments if you don't mind.

**Comment by mwalzer** [comment]
And just to complete a follow-up:
> Should the namespace coverage be sequence(7,4000000,4999999) instead?

Doesn't really matter once we're in PSI-MS where we don't have prefix exclusivity

> has_relation is a weird name for the typedef. It doesn't exist in PSI-MS-CV yet. Can we bring it in line with that CV or 
rename it to something better?

we should find a solution in the telco.

> has_type should be renamed to has_value_type for consistency with PSI-MS-CV.

Yep, I vaguely remember the last telco we agreed to that.

> has_tuple is not used and can be removed.

:+1:

> "ID free" etc. is a metric category imo, and should be specified using relationship rather than is_a.

We should find a solution in the telco. I think another agenda point would be the first bunch of terms. Maybe by category? Like that we can also introduce more structure as we go along. (#120) 

> MS:4000055 and MS:4000056: The name uses "quantile", but the definition details that it's quartiles.

See, that's the kind of review we need. We need to catch those. See PR #161

> A bunch of metrics are missing unit and type specifications that could be included.

A list would be great.

**Pull request merged**